FEXCore/Source/Interface/IR/Passes/x87StackOptimizationPass.cpp

```diff
@@ -0,0 +1,1892 @@
+  TopOffset = IREmit->_And(OpSize::i32Bit, TopOffset, GetConstant(7));
+  SetTopWithCache_Slow(TopOffset);
+}
+
+inline void X87StackOptimization::UpdateTop4Push_Slow() {
+  // Pop the top of the x87 stack
+  auto* TopOffset = GetTopWithCache_Slow();
+  TopOffset = IREmit->_Sub(OpSize::i32Bit, TopOffset, GetConstant(1));
+  TopOffset = IREmit->_And(OpSize::i32Bit, TopOffset, GetConstant(7));
+  SetTopWithCache_Slow(TopOffset);
+}
+
+// We synchronize stack values in a few occasions but one of the most important of those,
+// is when we move from fast to a slow path and need to make sure that the context is properly
+// written.
+Ref X87StackOptimization::SynchronizeStackValues() {
+  if (SlowPath) { // Nothing to do here.
+    LogMan::Msg::DFmt("Sync called in slow path - nothing to do here");
+    return GetTopWithCache_Slow();
+  }
+
+  // Store new top which is now the original top minus recorded top offset
+  // Careful with underflow wraparound.
+  const auto TopOffset = StackData.getTopOffset();
+  LogMan::Msg::DFmt("Writing stack to context - topoffset: {}", TopOffset);
+
+  if (TopOffset != 0) {
+    auto* OrigTop = GetTopWithCache_Slow();
+    Ref NewTop = nullptr;
+    if (TopOffset > 0) {
+      NewTop = IREmit->_And(OpSize::i32Bit, IREmit->_Sub(OpSize::i32Bit, OrigTop, GetConstant(TopOffset)), GetConstant(0x7));
+    } else {
+      NewTop = IREmit->_And(OpSize::i32Bit, IREmit->_Add(OpSize::i32Bit, OrigTop, GetConstant(-TopOffset)), GetConstant(0x7));
+    }
+    SetTopWithCache_Slow(NewTop);
+  }
+  StackData.setTopOffset(0);
+
+  // Before leaving we need to write the current values in the stack to
+  // context so that the values are correct. Copy SourceDataNode in the
+  // stack to the respective mmX register.
+  Ref TopValue = GetTopWithCache_Slow();
+  StackData.dump();
+  for (size_t i = 0; i < StackData.size(); ++i) {
+    const auto& [Valid, StackMember] = StackData.top(i);
+
+    if (Valid == StackSlot::UNUSED) {
+      continue;
+    }
+    Ref TopIndex = GetOffsetTopWithCache_Slow(i);
+    if (Valid == StackSlot::VALID) {
+      LogMan::Msg::DFmt("Writing StackData[{}]", i);
+      IREmit->_StoreContextIndexed(StackMember.StackDataNode, TopIndex, ReducedPrecisionMode ? 8 : 16, MMBaseOffset(), 16, FPRClass);
+    }
+  }
+  { // Set valid tags
+    uint8_t Mask = StackData.getValidMask();
+    Ref MaskC = GetConstant(Mask);
+    LogMan::Msg::DFmt("Writing valid tags 0x{:02x}", Mask);
+    if (Mask == 0xff) {
+      IREmit->_StoreContext(1, GPRClass, GetConstant(Mask), offsetof(FEXCore::Core::CPUState, AbridgedFTW));
+    } else if (Mask != 0) {
+      if (std::popcount(Mask) == 1) {
+        uint8_t BitIdx = __builtin_ctz(Mask);
+        SetX87ValidTag(GetOffsetTopWithCache_Slow(BitIdx), true);
+      } else {
+        // perform a rotate right on mask by top
+        // since we must operate on 32bits as a minimum:
+        // ror (mask, top) = ((M << 8) | M) >> TOP
```

Comment:
Any reason you can't _Ror?

Reply:
My ignorance maybe. :) 
Can we do an 8bit ror? I assumed it was only 32 or 64 bits rotate that were possible.

Reply:
Per discord - indeed we can't do an 8-bit ror. But we can use `_Orlshl` to cut an instruction.